### ISSUE TITLE
The new type hints cause pytest to hang after test session

### ISSUE THREAD

**Issue by KristobalJunta** [facelessuser/soupsieve#261]
Long story short: recently I faced a weird issue while running test for my project (that depends on soupsieve) - and it turned out to be caused by 422be13 for some reason that I can't figure out now.

```
Python           3.10.10
beautifulsoup4   4.10.0
soupsieve        2.4
```

My project depends on beautifulsoup4, which in turn depends on soupsieve.
After upgrading soupsieve to 2.4 I faced a problem - after running all the test suite, pytest was not completing. It shows test session summary (i.e. "X passed, Y skipped") and then just hangs - with no errors or output.

First thing I did was to pinpoint test case that caused this situation. It appeared to be roughly this (pseudo-code, actual code is private).

```python
import requests
from unittest.mock import MagicMock, patch

@patch("app.api_client.ping", MagicMock(return_value=True))
@patch("app.api_client.get_access_token", MagicMock(return_value=("token", "url")))
@patch("app.api_client.get_account_details", MagicMock(return_value={"id": "42", "email": "user@test.local"}))
def test_api_call(api_client):
    response = api_client.reconnect_sync()  # this makes a few API requests to 3d party and returns status
    assert response["data"]["status"] == "active"
```

Key point here appeared to be the `@patch`. When I removed the third one ("get_account_details") - the problem was gone.
I also tried to rewrite that patch into a context manager - and it helped too.
My project is relatively complex, and the issue also involves using graphene and DataLoaders. Removing dataloader invocation from the test case also fixes the problem.

So, my next step was to revert all project code to the "used to work, now doesn't" state and assume that it was caused by a change in some of the sub-dependencies.
tl;dr - it appeared to be soupsieve. And some `git bisect` investigation pointed me to commit 422be13 - which reworked type hints.
I am quite baffled at this point, and have no idea how these all factors came to cause the pytest bug.

Can confirm that soupsieve `2.3.2.post1` does not cause the issue.

Anyway, I will try to throw together a minimal reproduction of the bug - but would be glad if someone else has some insight on what may be happening here.

**Comment by KristobalJunta** [comment]
Sorry if I'm being not specific enough - but it was only today that I discovered the issue, and I'm still figuring out some parts of it myself.

**Comment by facelessuser** [comment]
The only type change thing that is different is that we are using `from __future__ import annotations` which treats all annotations as strings. That allows us to use things like `this | that` instead of `Union[this, that]` and `tuple[]` instead of `Tuple[]` on older Python versions. It allows me to use a more simple and more current syntax for mypy. You can look it over and see what the changes were.  The are extremely minor and should not impact runtime.

> Key point here appeared to be the `@patch`. When I removed the third one ("get_account_details") - the problem was gone.
> I also tried to rewrite that patch into a context manager - and it helped too.
> My project is relatively complex, and the issue also involves using graphene and DataLoaders. Removing dataloader invocation from the test case also fixes the problem.

This may only be tangentially related to what we are doing. Yes, there may be something that is weirdly exposing something here, but I am skeptical that there is something directly wrong with soupsieve, though I am more than happy to look into it further if more evidence can be shown to point to soupseive "doing something wrong". The actual changes in this release are fairly minor.

**Comment by KristobalJunta** [comment]
@facelessuser absolutely understand your point. soupseive was among the last things I would suspect in breaking the tests - but downgrading it actually removed the issue. But the code under testing does not directly invoke anything from the lib, which makes it weirder.
It seems to me, that there is some weird combination of factors that causes pytest misbehave on teardown. 
Will try and look further to boil it down to something more explainable.
